graphql_compiler/macros/macro_edge/helpers.py

```diff
@@ -53,6 +57,214 @@ def get_directives_for_ast(ast):
     return result
 
 
+def get_all_tag_names(ast):
+    """Return a set of strings containing tag names that appear in the query."""
+    return {
+        # Schema validation has ensured this exists
+        directive.arguments[0].value.value
+        for ast, directive in _yield_ast_nodes_with_directives(ast)
+        if directive.name.value == TagDirective.name
+    }
+
+
+def _remove_colocated_tags(non_macro_names, ast):
+    """Return an AST with at most one tag per field by removing tags.
+
+    Args:
+        non_macro_names: set of tag names that the user wrote. We prefer keeping these names.
+        ast: GraphQL query AST object that potentially has multiple colocated tags. This AST
```

Comment:
Same here - what is the exact type? Let's make it the first word.

Reply:
That's how we internally refer to them, rather than listing the six possible types. There's no single type that captures `GraphQL query AST object`

Reply:
Actually `Node` is that class, but I was just copying the style of our repo. Maybe it's worth going around the codebase and making the words `AST object` more precise, but I'd rather do it on master, not as part of the macro_system feature. https://github.com/graphql-python/graphql-core/blob/master/graphql/language/ast.py#L10

Reply:
Sounds good.